hibernate-core/src/main/java/org/hibernate/persister/collection/AbstractCollectionPersister.java

```diff
@@ -744,18 +744,21 @@ protected void logStaticSQL() {
 		}
 	}
 
-	private CollectionLoader standardCollectionLoader;
+	private volatile CollectionLoader standardCollectionLoader;
 
 	@Override
 	public void initialize(Object key, SharedSessionContractImplementor session) throws HibernateException {
 //		getAppropriateInitializer( key, session ).initialize( key, session );
 		determineLoaderToUse( key, session ).load( key, session );
 	}
 
+	// lazily initialize instance field via 'double-checked locking'
+	// see https://en.wikipedia.org/wiki/Double-checked_locking on why 'volatile' and local copy is used
 	protected CollectionLoader getStandardCollectionLoader() {
 		CollectionLoader localCopy = standardCollectionLoader;
 		if ( localCopy == null ) {
 			synchronized (this) {
+				localCopy = standardCollectionLoader;
 				if ( localCopy == null ) {
 					localCopy = createCollectionLoader( LoadQueryInfluencers.NONE );
 					standardCollectionLoader  = localCopy;
```

Comment:
see the explanation in screenshot to fully understand the tech details. It is subtle and error-prone.

![Screen Shot 2020-02-29 at 10 31 24 PM](https://user-images.githubusercontent.com/8211458/75618963-4feba080-5b43-11ea-8306-cb5bc803a365.png)